### ISSUE TITLE
Need another escape hatch for experimental

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty#16204]
@experimental definitions and experimental language imports can currently only used on code that is itself @experimental or that is compiled with a nightly version of the compiler.

This effectively locks experimental away in the poison cabinet. 

For instance, I see no practical version to use experimental features in the compiler itself. This is a shame. First, there is no issue in using experimental features in the compiler. If they change, we can change the compiler as well, it's the same repo. So the stability argument does not apply. Second, we forgo the large benefits of dog-fooding, which means that experimental feature 
are pushed out with much less testing and user experience than would have been the case otherwise. 

To change that, I propose we create an setting `-Yexperimental` which would make the compiler pretend it is a nightly version. I am not sure what this should mean for the Tasty version, however.

**Comment by bishabosha** [comment]
Perhaps we can check that the source file of the code is in a git repo matching the dotty repository?

**Comment by odersky** [comment]
That's one option. But it means we still put up very high hurdles for everyone else.

**Comment by smarter** [comment]
`-Yexperimental` could act as if all classes and traits were defined with `@experimental`.

**Comment by prolativ** [comment]
This idea seems familiar https://github.com/lampepfl/dotty/issues/14306#issuecomment-1069101894

**Comment by sjrd** [comment]
> part of metals depend on the compiler

As well as every compiler plugin out there, for example Scala Native.

**Comment by sjrd** [comment]
Also tooling such as Metals will not support these things. So if you *do* use them in the compiler, you cannot expect Metals to work for the compiler codebase anymore. Nor any IDE, for that matter.

**Comment by odersky** [comment]
> Also tooling such as Metals will not support these things. So if you do use them in the compiler, you cannot expect Metals to work for the compiler codebase anymore. Nor any IDE, for that matter.

Not necessarily. All I want is refer to some classes that are marked @experimental. I don't see why Metals would stop working on the compiler codebase that way. 

If we do introduce experimental language imports then it's different. It depends what they are. If they affect syntax it's possible that IDEs would not longer recognize this correctly. But that's not a complete show-stopper either. I have been working with occasional red underlines from metals for a long time previously.

If we label the compiler transitively as experimental then I agree it won't work. We need to have a way to override experimental rules locally without the codebase becoming itself experimental.

**Comment by odersky** [comment]
So, the radical option would indeed be to compile the compiler with -Yexperimental, which would force all plugins and other code using the compiler to be compiled with -Yexperimental as well. Actually, for metals it would only be those parts that rely on compiler internals, but it would take some work to factor them out cleany. 

In a sense that's just being honest. Internal compiler APIs are not stable, and can be changed at any time, which is our definition of what experimental is. Maybe that would push the compiler maintainers to develop richer stable APIs and other tools to use these APIs instead of compiler internals. But that's a process that requires biu in from many parties.

In summary I see the following options:

 - Add -Yexperimental in the sense that every compiled unit is treated as if it was annotated @experimental. That allows applications to use experimental features but it's not an attractive options for libraries or the compiler itself (since that one is also a library that users use).
 - Have a special exclude that allows the compiler to access experimental features without itself being experimental. The idea is that if an experimental feature changes, the compiler will change as well. Since everybody using the compiler needs to be prepared for any changes in the compiler, this is OK.
 - Make that special exclude also available to others in some way. 

What do you think? Which (combinations of) options should we go for?

**Comment by soronpo** [comment]
In any case, I think this should be brought up as a SIP for a committee vote.

**Comment by makkarpov** [comment]
> `-Yexperimental` could act as if all classes and traits were defined with `@experimental`.

`-Yskip:crossVersionChecks` could at as if `@experimental` was never existed.

**Comment by tgodzik** [comment]
> Have a special exclude that allows the compiler to access experimental features without itself being experimental. The idea is that if an experimental feature changes, the compiler will change as well. 

How does it work with tasty? Wouldn't it be a possibility that the tasty changes and breaks all the plugins downstream which compile with stable versions? We would need a tooling community build to test that out or Mima for Tasty?

> Since everybody using the compiler needs to be prepared for any changes in the compiler, this is OK.

Including all the tooling authors?  This works when we have a proper API to use, otherwise we should be careful about big changes that can impact tooling especially using experimental.

> This effectively locks experimental away in the poison cabinet.

Isn't it exactly what experimental is supposed to be? We don't want to use experimental in a real live code since it's something we are not yet sure about. And people might have less confidence in the compiler if it uses experimental features.

Overall, this change seems a bit hostile for tooling and I would try to keep in my mind that tooling authors can't be take for granted. Aside from Metals of course :sweat_smile:

**Comment by nicolasstucki** [comment]
@odersky I would like to understand what kind of experimental features we would like to access from the compiler.

I had the use case of accessing the `MacroAnnotation`. For this case, I just used plain JVM reflection. Then to make it nicer, I used `reflectiveSelectable` and created a mock interface that allowed me to call the experimental methods as if they were present.

```scala
import scala.reflect.Selectable.reflectiveSelectable
type MacroAnnotation = {
  def transform(using Quotes)(tree: Object/*Erased type of quotes.refelct.Definition*/): List[MemberDef /*quotes.refelct.Definition known to be MemberDef in QuotesImpl*/]
}
val annotInstance = ???.asInstanceOf[MacroAnnotation]
annotInstance.transform(using quotes)(tree)
```

Does this approach also work for your other use cases?

**Comment by odersky** [comment]
I think we should not jump through hoops with reflection here. We need to use something that's experimental, we should be able to get it without playing tricks like this.
